engine/src/main/java/com/agiletec/aps/system/services/keygenerator/KeyGeneratorManager.java

```diff
@@ -60,7 +60,10 @@ public void init() throws Exception {
      * nell'aggiornamento della chiave corrente.
      */
     public int getUniqueKeyCurrentValue() throws ApsSystemException {
-        int key = this.getCacheWrapper().getUniqueKeyCurrentValue() + 1;
+        int key;
+        synchronized (this) {

```

Comment:
This should be synchronized on a private member string variable. Synchronizing on `this` is not necessary and can cause issues down the line. For example, in the future there is a new method and somebody synchronizes on `this `then those two sections  are on the same lock and if one of them is long we have problems. Add a `private String UNIQUE_KEY_LOCK = "UNIQUE_KEY_LOCK" and synchronize on that.